pallene/ast.lua

```diff
@@ -88,18 +89,27 @@ declare_type("Field", {
 function ast.toplevel_names(tl_node)
     local names = {}
     local tag = tl_node._tag
-    if     tag == "ast.Toplevel.Func" then
-        table.insert(names, tl_node.decl.name)
-    elseif tag == "ast.Toplevel.Var" then
-        for _, decl in ipairs(tl_node.decls) do
-            table.insert(names, decl.name)
-        end
-    elseif tag == "ast.Toplevel.Typealias" then
+    if     tag == "ast.Toplevel.Typealias" then
         table.insert(names, tl_node.name)
     elseif tag == "ast.Toplevel.Record" then
         table.insert(names, tl_node.name)
     elseif tag == "ast.Toplevel.Builtin" then
         table.insert(names, tl_node.name)
+    elseif tag == "ast.Toplevel.Stat" then
+        local stat = tl_node.stat
+        if stat._tag == "ast.Stat.Decl" then
+            for _, decl in ipairs(tl_node.stat.decls) do
+                table.insert(names, decl.name)
+            end
+        elseif stat._tag == "ast.Stat.Func" then
+            local f_name = stat.name
+            local var = f_name.var
+            if var._tag ~= "ast.Var.Name" and var._tag ~= "ast.Var.Dot" then
+                error("impossible")
+            elseif var._tag == "ast.Var.Name" then
+                table.insert(names, var.name)
+            end
+        end
```

Comment:
I was using it to detect, for example, `local` functions and variables with the same name, or `typealiases` with the same name as `local`variables. If these duplicate names are being checked later in the code, then we really don't gain anything by keeping this function.

Reply:
It's OK if those things shadow, the same way as Lua. In fact, being more similar to Lua is one of the motivations for this rewrite...

The reason we currently have a no-shadowing rule is that Pallene to Lua translator works by adding a return statement to the end of the file:

    export function foo()
    end
    
    -- ...
    
becomes

   local function foo()
   end
   
   -- ...

    return { foo = foo }

However, if the `...` code has another definition of `foo`, it might shadow the definition we want to export, causing trouble. Anyway, the new syntax we're implementing should avoid this by making return statement explicit and getting rid of the magic.